bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BMapType.java

```diff
@@ -40,36 +41,32 @@
 @SuppressWarnings("unchecked")
 public class BMapType extends BType implements MapType {
 
-    private Type constraint;
+    private final Type constraint;
     private final boolean readonly;
     private IntersectionType immutableType;
 
+    public BMapType(Type constraint) {
+        this(constraint, false);
+    }
+
+    public BMapType(Type constraint, boolean readonly) {
+        this(TypeConstants.MAP_TNAME, constraint, null, readonly);
+    }
+
     /**
      * Create a type from the given name.
      *
-     * @param typeName string name of the type.
+     * @param typeName   string name of the type.
      * @param constraint constraint type which particular map is bound to.
-     * @param pkg package for the type.
+     * @param pkg        package for the type.
      */
     public BMapType(String typeName, Type constraint, Module pkg) {
-        super(typeName, pkg, MapValueImpl.class);
-        this.constraint = constraint;
-        this.readonly = false;
+        this(typeName, constraint, pkg, false);
     }
 
     public BMapType(String typeName, Type constraint, Module pkg, boolean readonly) {
         super(typeName, pkg, MapValueImpl.class);
-        this.constraint = constraint;
-        this.readonly = readonly;
-    }
-
-    public BMapType(Type constraint) {
-        this(constraint, false);
-    }
-
-    public BMapType(Type constraint, boolean readonly) {
-        super(TypeConstants.MAP_TNAME, null, MapValueImpl.class);
-        this.constraint = constraint;
+        this.constraint = readonly ? ReadOnlyUtils.getReadOnlyType(constraint) : constraint;
```

Comment:
IINM `io.ballerina.runtime.internal.values.ReadOnlyUtils#setImmutableIntersectionType`'s assumes the type passed is definitely a type that can be immutable and the default case seems to be assuming the type is a union. But when creating a type at runtime, I guess there could be scenarios where the user passes an invalid type? We may need to handle those with a proper error and include tests.

Reply:
Fixed this